Support multiple stability attributes on items

dianne opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/131824)*

## Motivation

Many unstable library items require the stabilization of multiple features before they can be stabilized. By allowing them to be annotated with multiple `#[unstable]` attributes, this prevents their accidental stabilization when some of those features are stabilized, and helps mitigate upgrade pains for unstable toolchain users.

## New stability attribute semantics

- If any `#[unstable]` attribute is present on an item, it is unstable. Likewise, any `#[rustc_const_unstable]` marks an item as const-unstable, and any `#[rustc_default_body_unstable]` marks it as default body-unstable.
- In order to use an unstable item, all feature gates specified by its `#[unstable]` attributes must be enabled. Conceptually, this is because it depends on multiple unstable features. Practically, this also means nightly toolchain users are less likely to need to replace one unstable feature gate with another in their crate-level attributes.
- Multiple `#[stable]` attributes may be present. The stabilization version of an item is the most recent stabilization version in present `#[stable]` attributes. Likewise, `#[stable]` attributes may be present on an unstable item. This simplifies macros that can apply stability attributes to items.

This is documented in rust-lang/rustc-dev-guide#2128.

## New syntactic constraints
- As a sanity check, an item is not allowed to have multiple stability attributes for the same feature.
- At most one `#[unstable]` attribute on an item may have a `reason` provided (and likewise for `#[rustc_const_unstable]` and `#[rustc_default_body_unstable]`). At a glance, the `reason` meta-item was used to describe the reason for an item being unstable, rather than an individual feature it depends on. Enforcing this makes the diagnostic formatting much cleaner.
- If any `#[unstable]` attribute on an item has a `soft` marker, it must be present on all `#[unstable]` attributes on that item. Items can't be partially soft-unstable, and this helps prevent accidentally making a soft-unstable item unusable on stable when stabilizing one of the features it requires.

Fixes rust-lang/rust#94770

Based on rust-lang/rust#94988

~~PR for a diagnostic formatting change this uses: rust-lang/rust#132544~~

Some notes on assumptions I've made and how I've handled them:
- Stability structs are specialized to how they're used for stability checking and rustdoc. `rustc_passes::lib_features` also reads stability attributes, but it parses them itself. I considered it out of the scope of this PR refactor that too, but once attribute handling is reworked to be more unified in the future, this may need further adjustment (but hopefully not too much).
- Under the assumption that most library items will still have at most one `#[unstable]` attribute, this uses `SmallVec<[_; 1]>` for storage. Is this assumption reasonable? I haven't done any performance testing.
- Since `StabilityLevel` no longer has `Copy`, stability structs are now arena-allocated; that was the simplest fix. I'm not sure what the performance implications are. If it's an issue, I can try try making it only allocate for multiple unstable features (which would also let `StabilityLevel` regain `Copy`).
- ~~As a stopgap until rust-lang/rust#132541 lands, this is using a `ConstStabilityLevel` enum for const-stability levels. Once that's merged, I should be able to move const-stability levels back to using `StabilityLevel`.~~

This doesn't update any libraries to add additional unstable attributes to items that should have them.

dianne committed:
Refactor stability structs

This moves stability structs' `feature` fields into
`StabilityLevel::Unstable` and `ConstStabilityLevel::Unstable`, in
preparation to support multiple unstable attributes on items.

Seemingly, the `feature` field isn't used with the
`StabilityLevel::Stable` variant, so I haven't included it.
`rustc_passes::lib_features` uses the 'feature' meta-item for 'stable'
attributes, but it extracts them itself, rather than relying on
`rustc_attr`.

dianne committed:
factor out some commonalities in the `find_stability` family of functions

the logic for adding unstable attrs gets a bit messier when supporting multiple
instances thereof. this keeps that from being duplicated in 3 places.

dianne committed:
compiler: support multiple stability attributes on items

This changes the text for E0544. Unfortunately, the example doesn't make
much sense anymore.

The way this merges stability levels together is made to work for
stability-checking and rustdoc; as far as I can tell, only
`rustc_passes::lib_features` needs them separate, and it extracts them
itself.

This includes the clarified/fixed const-unstable semantics that were
previously in a different commit.

dianne committed:
rustdoc: support multiple stability attributes

dianne committed:
rustdoc: report all unstable features needed for items

dianne committed:
Make unstable library feature errors translatable

Translation is in an awkward position, but this provides a similar
interface for both errors and soft-unstable lints, and enables the use
of `DiagSymbolList` for simplifying error message construction.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11382184447/job/31665174592) [(plain)](https://github.com/rust-lang/rust/commit/ebf78525bf0ebdb38436a9ed700ec20da352312f/checks/31665174592/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.984 Building wheels for collected packages: reuse
#16 2.985   Building wheel for reuse (pyproject.toml): started
#16 3.232   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.233   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#16 3.233   Stored in directory: /tmp/pip-ephem-wheel-cache-2p76vwku/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.236 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.634 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.635 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.7s
---
 Documenting rustc_hir_pretty v0.0.0 (/checkout/compiler/rustc_hir_pretty)
error: unresolved link to `rustc_lint_defs::builtin::SOFT_UNSTABLE`
   --> compiler/rustc_middle/src/error.rs:186:27
    |
186 | /// Lint diagnostic for [`rustc_lint_defs::builtin::SOFT_UNSTABLE`]
    |
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`


 Documenting rustc_lint_defs v0.0.0 (/checkout/compiler/rustc_lint_defs)
 Documenting rustc_llvm v0.0.0 (/checkout/compiler/rustc_llvm)
error: could not document `rustc_middle`
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Thu Oct 17 09:54:48 UTC 2024
  network time: Thu, 17 Oct 2024 09:54:48 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #132020) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dianne committed:
Add test for mixed stability levels

bors commented:
:umbrella: The latest upstream changes (presumably #132027) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #131985) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #131349) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dianne committed:
clippy: update `is_stable_const_fn` helper for multiple const-stability attrs

dianne commented:
cc @RalfJung @compiler-errors This rebase was pretty substantial and required some minor refactors. Would either of you mind giving the const-stability parts a look to make sure they make sense and don't go against planned future changes?

cc @rust-lang/libs-api I realize I should probably ping here as well. Does the design supporting a mix of unstable/stable attributes sound good? I'll try documenting it in the dev guide and std dev guide once that's finalized.

bors commented:
:umbrella: The latest upstream changes (presumably #131284) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

RalfJung commented:
The PR should explain what the semantics are of having multiple `unstable` attributes. Do *all* gates need to be active to use the feature, or only some?

> This supports multiple #[stable] attributes, and a mix of #[stable] and #[unstable], in case that's helpful for tracking which features library items depended on as they stabilize (and because it's less difficult to reword E0544 this way).

That sounds really strange. What is the semantics of this? And what is the point?

RalfJung commented on compiler/rustc_const_eval/src/check_consts/check.rs:
```diff
@@ -773,39 +789,52 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
+                                .filter(|u| !is_feature_enabled(u))
+                                .map(|u| u.into())
+                                .collect()
+                        };
```
This looks like it changes the logic -- now the closure defined up here also gets used below.

I'd prefer if you didn't change the logic but just did the "obvious" variant to generalize to multiple gates. Otherwise this will have to go somewhere on my pretty big "needs more thorough review" pile.

dianne replied:
I think I've fixed this. It should be the obvious generalization of the previous logic now.

RalfJung replied:
Yes it looks quite clear now, thanks. :)

RalfJung commented on compiler/rustc_attr/src/builtin.rs:
```diff
@@ -161,6 +151,20 @@ pub enum StabilityLevel {
 }
 
+/// The available const-stability levels for const functions.
+/// For details see [#131349](https://github.com/rust-lang/rust/pull/131349).
```
A PR is not a good reference for this as its description will inevitably go out of date.

dianne replied:
I've changed the link to be to the dev guide and moved it to the `ConstStability` struct. Though it doesn't document the internal representation, I think having the link to the description of recursive const-stability and `const_stable_indirect` on the struct is helpful.

RalfJung commented on compiler/rustc_attr/src/builtin.rs:
```diff
@@ -161,6 +151,20 @@ pub enum StabilityLevel {
+    Unstable { unstables: SmallVec<[Unstability; 1]> },
+    /// For functions with no explicit const-stability attribute that require checking recursive
+    /// const stability. This is either an unmarked const fn or a `const_stable_indirect` intrinsic.
+    Implicit,
```
Please rename this to `ImplicitUnstable`, because we never want anything to be "implicitly stable".

Also FWIW this variant will disappear when I have the time to continue working on the const stability checks.

dianne replied:
Done. Do you think this should be blocked on that being merged?

RalfJung replied:
My PR is now up: https://github.com/rust-lang/rust/pull/132541

RalfJung commented on tests/ui/consts/const-unstable-intrinsic.stderr:
```diff
@@ -50,7 +50,10 @@ error: `min_align_of_val` is not yet stable as a const intrinsic
 LL |         unstable_intrinsic::old_way::min_align_of_val(&x);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
+   = note: use of unstable library feature `unstable`
```
Why does the note repeat the error description?

The diagnostic just above doesn't do this.

dianne commented:
> The PR should explain what the semantics are of having multiple `unstable` attributes. Do _all_ gates need to be active to use the feature, or only some?
> 
> > This supports multiple #[stable] attributes, and a mix of #[stable] and #[unstable], in case that's helpful for tracking which features library items depended on as they stabilize (and because it's less difficult to reword E0544 this way).
> 
> That sounds really strange. What is the semantics of this? And what is the point?

I've updated the PR description to address this. To your second point, I'm working off of [this specification](https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/making.20.60BuiltinLintDiag.3A.3AUnstableFeature.60.20translatable/near/474755755). Treating it as a matter of preference, I decided to implement it both with and without support for multiple `#[stable]` attributes in separate commits so that I'd have an implementation to point to before asking which approach is nicer. I'll work on addressing your review comments shortly. Thank you for giving this a look!

dianne committed:
Add sanity checks for `reason` and `soft` on unstable attributes

`reason` must appear at most once: it's the reason for the item being unstable, rather than a
particular feature. This simplifies diagnostic formatting.

`soft` must either be on all or no unstable attributes: it doesn't make sense for something to be
partially soft, and allowing inconsistent softness markers would risk an item accidentally becoming
properly unstable as its features stabilize.

RalfJung commented:
The updated PR description is great, thanks. :)

Personally I don't think multiple stable attributes, or both stable and unstable, are worth it. But let's see what t-libs-api thinks.

bors commented:
:umbrella: The latest upstream changes (presumably #132145) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #132435) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #132470) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #132513) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

RalfJung commented:
I would suggest you make a separate PR for the change in diagnostics (`'` -> `` ` ``); that one can land ahead and that should drastically reduce the amount of conflicts you see here.

dianne commented:
Good idea! That PR is now up: #132544

RalfJung commented:
That will also make reviewing this PR a lot easier by reducing the noise in the diff. :)

bors commented:
:umbrella: The latest upstream changes (presumably #132603) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #132626) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Amanieu commented:
We discussed this in the libs-api meeting today and we're happy with this as proposed. Support for multiple `stable` attribute was somewhat controversial but in the end we felt that it would be useful in macros which apply a mix of stability attributes to many items.

dianne commented:
> Whoever happens to merge their changes first will force the other to retrofit their changes onto it haha, but with some luck for you, that'll be me who has to do the retrofitting :3

I've been wondering about this. My one hope if I do the rebase is that the attribute rework will make this cleaner in the end. :sweat_smile: 


> We discussed this in the libs-api meeting today and we're happy with this as proposed.

Nice! I'll get to work on documenting it soon.

dianne commented:
Is that a new notification? Unless I made a mistake, I *think* all I did was resolve the merge conflict straightforwardly, so see the above discussion for how this touches the CTFE machinery.

RalfJung commented:
We changed the notification settings recently and apparently that triggers a new notification.

bors commented:
:umbrella: The latest upstream changes (presumably #132831) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dianne committed:
Allow multiple unstable features in diagnostics

dianne commented:
I did some cleanup to the diagnostic formatting and commit history. Now only the tests for stability attribute sanity (and the tests I added) have diffs. Hopefully it will be more digestible now, and easier to split apart if needed.

bors commented:
:umbrella: The latest upstream changes (presumably #132954) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
> The updated PR description is great, thanks. :)
> 
> Personally I don't think multiple stable attributes, or both stable and unstable, are worth it. But let's see what t-libs-api thinks.

I agree that the cost/benefit ratio here is too high, from the compiler maintenance point of view.
It's not even a user facing feature, just a minor convenience for the library people.

bors commented:
:umbrella: The latest upstream changes (presumably #132910) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #133179) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@dianne i know this is waiting for review, but this has gathered conflicts so it would be better to rebase it and we can get it reviewed after that. Thanks

jdonszelmann commented:
I'm terribly sorry but I'm afraid the conflicts are rather complicated because the way attributes work itself was partially changed. Let me know if I need to help out with info about the new system.

dianne commented:
This was really conflict-prone anyway, so I've enjoyed the break from maintaining it. ^^ When I rebase/rewrite it, I'll see how easy it'd be to split the const-stability changes into a separate PR; the const-eval parts were the most conflict-heavy, and that that should hopefully make this a bit more digestible too. I'll also see about once again starting off with only supporting multiple unstable attributes.

@jdonszelmann ty! I'll let you know if I have any specific questions. Do you expect any more reworks on your end to how stability attributes are handled? No worries either way! I'm not in a rush at least, so I'm happy to wait for the dust to settle if that means fewer future conflicts.

alex-semenyuk commented:
@dianne
Thanks for your contribution
From wg-triage. Any updates on this PR?

dianne commented:
None yet. I'm planning on getting back to this in a week or two and paring it down as much as I can.

alex-semenyuk commented:
@dianne 
Are you waiting answer from @jdonszelmann?
Meantime, please rebase

dianne commented:
The rebase requires a substantial amount of rewriting, so since making `Stability` non-`Copy` is a fairly messy change, I was hoping to figure out the path forward for this PR before then (ideally with `Stability` still being `Copy`), hence the ask. I'll convert the PR to a draft for now, since I'd prefer not to maintain it without a clearer idea of how it should look in the end. If it would be preferable (or if someone else would like to tackle this), I can close this PR instead and release my assignment.

Dylan-DPC commented:
Closing this as the author wants to tackle this when they have a better idea of the solution and this pr is heavily bitrotty.
